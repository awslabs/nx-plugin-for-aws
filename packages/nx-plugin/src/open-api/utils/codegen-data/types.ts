/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import type { Plugin } from '@hey-api/openapi-ts';
import type { OpenAPIV3, OpenAPIV3_1 } from 'openapi-types';

export type ClientData = Parameters<Plugin.LegacyHandler<any>>[0]['client'];
export type Operation = ClientData['services'][number]['operations'][number];
export type Model = ClientData['models'][number];

export interface CodeGenData extends ClientData {
  className: string;
  info: OpenAPIV3.InfoObject | OpenAPIV3_1.InfoObject;
  allOperations: Operation[];
  operationsByTag: { [tag: string]: Operation[] };
  untaggedOperations: Operation[];
  vendorExtensions: { [key: string]: any };
}

export const flattenModelLink = (link?: Model | Model[]): Model =>
  link === undefined ? undefined : Array.isArray(link) ? link[0] : link;

// Model types which indicate it is composed (ie inherits/mixin's another schema)
export const COMPOSED_SCHEMA_TYPES = new Set(['one-of', 'any-of', 'all-of']);
export const COLLECTION_TYPES = new Set(['array', 'dictionary']);
export const PRIMITIVE_TYPES = new Set([
  'string',
  'integer',
  'number',
  'boolean',
  'null',
  'any',
  'binary',
  'void',
]);
