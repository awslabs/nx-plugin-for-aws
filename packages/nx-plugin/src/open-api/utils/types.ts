/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import type { OpenAPIV3, OpenAPIV3_1 } from 'openapi-types';

export type Spec = OpenAPIV3.Document | OpenAPIV3_1.Document;
