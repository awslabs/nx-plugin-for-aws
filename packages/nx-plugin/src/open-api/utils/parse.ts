/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import SwaggerParser from '@apidevtools/swagger-parser';
import { Spec } from './types';

const TREE_PROTOCOL_PREFIX = 'workspace://';

/**
 * Parse an OpenAPI specification given its path within the tree
 */
export const parseOpenApiSpec = async (
  tree: Tree,
  specPath: string,
): Promise<Spec> => {
  // Use a custom protocol prefix to ensure that the swagger parser doesn't try and read from the filesystem itself
  const spec = await SwaggerParser.bundle(
    `${TREE_PROTOCOL_PREFIX}/${specPath}`,
    {
      resolve: {
        // Use a custom resolver which reads files from the tree rather than directly from the filesystem
        file: {
          order: 1,
          read: (file) => {
            const path = file.url.slice(TREE_PROTOCOL_PREFIX.length);
            if (!tree.exists(path)) {
              const message = `Unable to find ${specPath} in workspace ${tree.root}`;

              // Write an error to console as swagger parser's error message may be confusing with our custom protocol prefix
              console.error(message);
              throw new Error(message);
            }
            return tree.read(path);
          },
          canRead: (file) => file.url.startsWith(TREE_PROTOCOL_PREFIX),
        },
        http: false,
      },
    },
  );

  if ('swagger' in spec) {
    throw new Error(
      `OpenAPI v2 specifications are not supported. Please use OpenAPI v3`,
    );
  }
  return spec;
};
