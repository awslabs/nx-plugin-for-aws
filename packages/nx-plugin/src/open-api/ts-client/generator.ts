/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { generateFiles, Tree } from '@nx/devkit';
import { OpenApiTsClientGeneratorSchema } from './schema';
import { parseOpenApiSpec } from '../utils/parse';
import { buildOpenApiCodeGenData } from '../utils/codegen-data';
import * as path from 'path';
import { CodeGenData } from '../utils/codegen-data/types';
import { formatFilesInSubtree } from '../../utils/format';

/**
 * Generates typescript client from an openapi spec
 */
export const openApiTsClientGenerator = async (
  tree: Tree,
  options: OpenApiTsClientGeneratorSchema,
) => {
  const data = await buildOpenApiCodeGenerationData(
    tree,
    options.openApiSpecPath,
  );

  generateOpenApiTsClient(tree, data, options.outputPath);

  await formatFilesInSubtree(tree);
};

/**
 * Build a data structure which can be used to generate code from OpenAPI
 */
export const buildOpenApiCodeGenerationData = async (
  tree: Tree,
  specPath: string,
) => {
  const spec = await parseOpenApiSpec(tree, specPath);
  return await buildOpenApiCodeGenData(spec);
};

/**
 * Generate an OpenAPI typescript client in the target directory
 */
export const generateOpenApiTsClient = (
  tree: Tree,
  data: CodeGenData,
  outputPath: string,
) => {
  generateFiles(tree, path.join(__dirname, 'files'), outputPath, data);
};

export default openApiTsClientGenerator;
