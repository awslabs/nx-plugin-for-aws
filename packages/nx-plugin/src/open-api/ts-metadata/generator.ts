/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { generateFiles, Tree } from '@nx/devkit';
import { OpenApiMetadataSchema } from './schema';
import { parseOpenApiSpec } from '../utils/parse';
import { buildOpenApiCodeGenData } from '../utils/codegen-data';
import { formatFilesInSubtree } from '../../utils/format';
import path from 'path';
import { CodeGenData } from '../utils/codegen-data/types';

/**
 * Generates typescript metadata from an openapi spec
 */
export const openApiTsMetadataGenerator = async (
  tree: Tree,
  options: OpenApiMetadataSchema,
) => {
  const spec = await parseOpenApiSpec(tree, options.openApiSpecPath);

  const data = await buildOpenApiCodeGenData(spec);

  generateOpenApiTsMetadata(tree, data, options.outputPath);

  await formatFilesInSubtree(tree);
};

/**
 * Generate OpenAPI typescript metadata in the target directory
 */
export const generateOpenApiTsMetadata = (
  tree: Tree,
  data: CodeGenData,
  outputPath: string,
) => {
  generateFiles(tree, path.join(__dirname, 'files'), outputPath, data);
};

export default openApiTsMetadataGenerator;
