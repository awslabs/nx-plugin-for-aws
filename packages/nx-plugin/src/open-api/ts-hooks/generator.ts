/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import { OpenApiHooksSchema } from './schema';
import { parseOpenApiSpec } from '../utils/parse';
import { buildOpenApiCodeGenData } from '../utils/codegen-data';
import { generateOpenApiTsClient } from '../ts-client/generator';

/**
 * Generates typescript hooks from an openapi spec
 */
export const openApiTsHooksGenerator = async (
  tree: Tree,
  options: OpenApiHooksSchema,
) => {
  const spec = await parseOpenApiSpec(tree, options.openApiSpecPath);

  const data = await buildOpenApiCodeGenData(spec);

  generateOpenApiTsClient(tree, data, options.outputPath);

  // TODO: generate hooks which wrap the client
};

export default openApiTsHooksGenerator;
