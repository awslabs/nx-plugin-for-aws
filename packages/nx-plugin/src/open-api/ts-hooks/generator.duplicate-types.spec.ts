/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import { createTreeUsingTsSolutionSetup } from '../../utils/test';
import { Spec } from '../utils/types';
import { openApiTsHooksGenerator } from './generator';
import { TypeScriptVerifier } from '../../utils/test/ts.spec';

describe('openApiTsHooksGenerator - Duplicate Types', () => {
  let tree: Tree;
  const title = 'TestApi';
  const verifier = new TypeScriptVerifier(['@tanstack/react-query']);

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
  });

  const validateTypeScript = (paths: string[]) => {
    verifier.expectTypeScriptToCompile(tree, paths);
  };

  it('should handle duplicate type names when operation ID matches schema name', async () => {
    const spec: Spec = {
      openapi: '3.0.0',
      info: { title, version: '1.0.0' },
      components: {
        schemas: {
          // This schema conflicts with the request type that would be generated for the 'chat' operation
          ChatRequest: {
            type: 'object',
            properties: {
              message: { type: 'string' },
            },
            required: ['message'],
          },
        },
      },
      paths: {
        '/chat': {
          post: {
            operationId: 'chat',
            description: 'Send a chat message',
            requestBody: {
              required: true,
              content: {
                'application/json': {
                  schema: {
                    $ref: '#/components/schemas/ChatRequest',
                  },
                },
              },
            },
            responses: {
              '200': {
                description: 'Chat response',
                content: {
                  'application/json': {
                    schema: {
                      type: 'object',
                      properties: {
                        reply: { type: 'string' },
                      },
                      required: ['reply'],
                    },
                  },
                },
              },
            },
          },
        },
      },
    };

    tree.write('openapi.json', JSON.stringify(spec));

    await openApiTsHooksGenerator(tree, {
      openApiSpecPath: 'openapi.json',
      outputPath: 'src/generated',
    });

    validateTypeScript([
      'src/generated/client.gen.ts',
      'src/generated/types.gen.ts',
      'src/generated/options-proxy.gen.ts',
    ]);

    const typesContent = tree.read('src/generated/types.gen.ts', 'utf-8');
    const optionsProxyContent = tree.read(
      'src/generated/options-proxy.gen.ts',
      'utf-8',
    );

    // Verify that the original schema is preserved
    expect(typesContent).toContain('export type ChatRequest = {');
    expect(typesContent).toContain('message: string;');

    // Verify that the operation request type is renamed to avoid conflict
    expect(typesContent).toContain(
      'export type ChatOperationRequest = ChatRequest;',
    );

    // Verify that the options proxy imports the correct type
    expect(optionsProxyContent).toContain('ChatOperationRequest,');

    // Verify the generated options proxy content
    expect(optionsProxyContent).toMatchSnapshot();
  });

  it('should handle multiple duplicate type names correctly', async () => {
    const spec: Spec = {
      openapi: '3.0.0',
      info: { title, version: '1.0.0' },
      components: {
        schemas: {
          // Multiple conflicting schemas
          UserRequest: {
            type: 'object',
            properties: {
              id: { type: 'string' },
            },
            required: ['id'],
          },
          OrderRequest: {
            type: 'object',
            properties: {
              productId: { type: 'string' },
            },
            required: ['productId'],
          },
        },
      },
      paths: {
        '/users': {
          post: {
            operationId: 'user',
            description: 'Create a user',
            requestBody: {
              required: true,
              content: {
                'application/json': {
                  schema: {
                    $ref: '#/components/schemas/UserRequest',
                  },
                },
              },
            },
            responses: {
              '201': {
                description: 'User created',
                content: {
                  'application/json': {
                    schema: {
                      type: 'object',
                      properties: {
                        id: { type: 'string' },
                        name: { type: 'string' },
                      },
                      required: ['id', 'name'],
                    },
                  },
                },
              },
            },
          },
        },
        '/orders': {
          post: {
            operationId: 'order',
            description: 'Create an order',
            requestBody: {
              required: true,
              content: {
                'application/json': {
                  schema: {
                    $ref: '#/components/schemas/OrderRequest',
                  },
                },
              },
            },
            responses: {
              '201': {
                description: 'Order created',
                content: {
                  'application/json': {
                    schema: {
                      type: 'object',
                      properties: {
                        id: { type: 'string' },
                        total: { type: 'number' },
                      },
                      required: ['id', 'total'],
                    },
                  },
                },
              },
            },
          },
        },
      },
    };

    tree.write('openapi.json', JSON.stringify(spec));

    await openApiTsHooksGenerator(tree, {
      openApiSpecPath: 'openapi.json',
      outputPath: 'src/generated',
    });

    validateTypeScript([
      'src/generated/client.gen.ts',
      'src/generated/types.gen.ts',
      'src/generated/options-proxy.gen.ts',
    ]);

    const typesContent = tree.read('src/generated/types.gen.ts', 'utf-8');
    const optionsProxyContent = tree.read(
      'src/generated/options-proxy.gen.ts',
      'utf-8',
    );

    // Verify that the original schemas are preserved
    expect(typesContent).toContain('export type UserRequest = {');
    expect(typesContent).toContain('export type OrderRequest = {');

    // Verify that both operation request types are renamed to avoid conflicts
    expect(typesContent).toContain(
      'export type UserOperationRequest = UserRequest;',
    );
    expect(typesContent).toContain(
      'export type OrderOperationRequest = OrderRequest;',
    );

    // Verify that the options proxy imports the correct types
    expect(optionsProxyContent).toContain('UserOperationRequest,');
    expect(optionsProxyContent).toContain('OrderOperationRequest,');

    // Verify the generated options proxy content
    expect(optionsProxyContent).toMatchSnapshot();
  });

  it('should not rename when there are no conflicts', async () => {
    const spec: Spec = {
      openapi: '3.0.0',
      info: { title, version: '1.0.0' },
      components: {
        schemas: {
          // This schema does NOT conflict with any operation request type
          UserData: {
            type: 'object',
            properties: {
              name: { type: 'string' },
            },
            required: ['name'],
          },
        },
      },
      paths: {
        '/users': {
          post: {
            operationId: 'createUser',
            description: 'Create a user',
            requestBody: {
              required: true,
              content: {
                'application/json': {
                  schema: {
                    $ref: '#/components/schemas/UserData',
                  },
                },
              },
            },
            responses: {
              '201': {
                description: 'User created',
                content: {
                  'application/json': {
                    schema: {
                      type: 'object',
                      properties: {
                        id: { type: 'string' },
                      },
                      required: ['id'],
                    },
                  },
                },
              },
            },
          },
        },
      },
    };

    tree.write('openapi.json', JSON.stringify(spec));

    await openApiTsHooksGenerator(tree, {
      openApiSpecPath: 'openapi.json',
      outputPath: 'src/generated',
    });

    validateTypeScript([
      'src/generated/client.gen.ts',
      'src/generated/types.gen.ts',
      'src/generated/options-proxy.gen.ts',
    ]);

    const typesContent = tree.read('src/generated/types.gen.ts', 'utf-8');
    const optionsProxyContent = tree.read(
      'src/generated/options-proxy.gen.ts',
      'utf-8',
    );

    // Verify that the original schema is preserved
    expect(typesContent).toContain('export type UserData = {');

    // Verify that the operation request type uses the standard naming (no conflict)
    expect(typesContent).toContain('export type CreateUserRequest = UserData;');

    // Verify that the options proxy imports the standard type name
    expect(optionsProxyContent).toContain('CreateUserRequest,');

    // Should NOT contain the OperationRequest suffix since there's no conflict
    expect(typesContent).not.toContain('CreateUserOperationRequest');

    // Verify the generated options proxy content
    expect(optionsProxyContent).toMatchSnapshot();
  });

  it('should handle query operations with duplicate type names', async () => {
    const spec: Spec = {
      openapi: '3.0.0',
      info: { title, version: '1.0.0' },
      components: {
        schemas: {
          // This schema conflicts with the request type that would be generated for the 'search' operation
          SearchRequest: {
            type: 'object',
            properties: {
              query: { type: 'string' },
            },
            required: ['query'],
          },
        },
      },
      paths: {
        '/search': {
          get: {
            operationId: 'search',
            description: 'Search for items',
            parameters: [
              {
                name: 'query',
                in: 'query',
                required: true,
                schema: { type: 'string' },
              },
              {
                name: 'limit',
                in: 'query',
                required: false,
                schema: { type: 'integer', default: 10 },
              },
            ],
            responses: {
              '200': {
                description: 'Search results',
                content: {
                  'application/json': {
                    schema: {
                      type: 'object',
                      properties: {
                        results: {
                          type: 'array',
                          items: { type: 'string' },
                        },
                      },
                      required: ['results'],
                    },
                  },
                },
              },
            },
          },
        },
      },
    };

    tree.write('openapi.json', JSON.stringify(spec));

    await openApiTsHooksGenerator(tree, {
      openApiSpecPath: 'openapi.json',
      outputPath: 'src/generated',
    });

    validateTypeScript([
      'src/generated/client.gen.ts',
      'src/generated/types.gen.ts',
      'src/generated/options-proxy.gen.ts',
    ]);

    const typesContent = tree.read('src/generated/types.gen.ts', 'utf-8');
    const optionsProxyContent = tree.read(
      'src/generated/options-proxy.gen.ts',
      'utf-8',
    );

    // Verify that the original schema is preserved
    expect(typesContent).toContain('export type SearchRequest = {');
    expect(typesContent).toContain('query: string;');

    // Verify that the operation request type is renamed to avoid conflict
    expect(typesContent).toContain(
      'export type SearchOperationRequest = SearchRequestQueryParameters',
    );

    // Verify that the options proxy imports the correct type
    expect(optionsProxyContent).toContain('SearchOperationRequest,');

    // Verify the generated options proxy content
    expect(optionsProxyContent).toMatchSnapshot();
  });

  it('should handle mutation operations with duplicate type names', async () => {
    const spec: Spec = {
      openapi: '3.0.0',
      info: { title, version: '1.0.0' },
      components: {
        schemas: {
          // This schema conflicts with the request type that would be generated for the 'update' operation
          UpdateRequest: {
            type: 'object',
            properties: {
              id: { type: 'string' },
              data: { type: 'object' },
            },
            required: ['id', 'data'],
          },
        },
      },
      paths: {
        '/items/{id}': {
          put: {
            operationId: 'update',
            description: 'Update an item',
            parameters: [
              {
                name: 'id',
                in: 'path',
                required: true,
                schema: { type: 'string' },
              },
            ],
            requestBody: {
              required: true,
              content: {
                'application/json': {
                  schema: {
                    $ref: '#/components/schemas/UpdateRequest',
                  },
                },
              },
            },
            responses: {
              '200': {
                description: 'Item updated',
                content: {
                  'application/json': {
                    schema: {
                      type: 'object',
                      properties: {
                        success: { type: 'boolean' },
                      },
                      required: ['success'],
                    },
                  },
                },
              },
            },
          },
        },
      },
    };

    tree.write('openapi.json', JSON.stringify(spec));

    await openApiTsHooksGenerator(tree, {
      openApiSpecPath: 'openapi.json',
      outputPath: 'src/generated',
    });

    validateTypeScript([
      'src/generated/client.gen.ts',
      'src/generated/types.gen.ts',
      'src/generated/options-proxy.gen.ts',
    ]);

    const typesContent = tree.read('src/generated/types.gen.ts', 'utf-8');
    const optionsProxyContent = tree.read(
      'src/generated/options-proxy.gen.ts',
      'utf-8',
    );

    // Verify that the original schema is preserved
    expect(typesContent).toContain('export type UpdateRequest = {');
    expect(typesContent).toContain('id: string;');
    expect(typesContent).toContain('data: { [key: string]: unknown };');

    // Verify that the operation request type is renamed to avoid conflict
    expect(typesContent).toContain(
      'export type UpdateOperationRequest = UpdateRequestPathParameters &',
    );

    // Verify that the options proxy imports the correct type
    expect(optionsProxyContent).toContain('UpdateOperationRequest,');

    // Verify the generated options proxy content
    expect(optionsProxyContent).toMatchSnapshot();
  });
});
