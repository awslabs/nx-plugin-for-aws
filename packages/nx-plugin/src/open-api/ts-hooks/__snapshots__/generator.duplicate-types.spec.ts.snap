// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`openApiTsHooksGenerator - Duplicate Types > should handle duplicate type names when operation ID matches schema name 1`] = `
"import type { UseMutationOptions } from '@tanstack/react-query';
import type {
  ChatOperationRequest,
  ChatError,
  Chat200Response,
} from './types.gen.js';
import { TestApi } from './client.gen.js';

export interface TestApiOptionsProxyConfig {
  client: TestApi;
}

export class TestApiOptionsProxy {
  private $client: TestApi;

  constructor({ client }: TestApiOptionsProxyConfig) {
    this.$client = client;

    this._chatMutationKey = this._chatMutationKey.bind(this);
    this._chatMutationOptions = this._chatMutationOptions.bind(this);
  }

  public queryKey = () => ['TestApi'];

  private _chatMutationKey() {
    return [...this.queryKey(), 'chat'];
  }
  private _chatMutationOptions(
    options?: UseMutationOptions<
      Chat200Response,
      ChatError,
      ChatOperationRequest
    >,
  ): UseMutationOptions<Chat200Response, ChatError, ChatOperationRequest> {
    return {
      mutationFn: (input) => this.$client.chat(input),
      mutationKey: this._chatMutationKey(),
      ...options,
    };
  }

  public chat = {
    mutationKey: this._chatMutationKey.bind(this),
    mutationOptions: this._chatMutationOptions.bind(this),
  };
}
"
`;

exports[`openApiTsHooksGenerator - Duplicate Types > should handle multiple duplicate type names correctly 1`] = `
"import type { UseMutationOptions } from '@tanstack/react-query';
import type {
  OrderOperationRequest,
  UserOperationRequest,
  OrderError,
  UserError,
  Order201Response,
  User201Response,
} from './types.gen.js';
import { TestApi } from './client.gen.js';

export interface TestApiOptionsProxyConfig {
  client: TestApi;
}

export class TestApiOptionsProxy {
  private $client: TestApi;

  constructor({ client }: TestApiOptionsProxyConfig) {
    this.$client = client;

    this._orderMutationKey = this._orderMutationKey.bind(this);
    this._orderMutationOptions = this._orderMutationOptions.bind(this);
    this._userMutationKey = this._userMutationKey.bind(this);
    this._userMutationOptions = this._userMutationOptions.bind(this);
  }

  public queryKey = () => ['TestApi'];

  private _orderMutationKey() {
    return [...this.queryKey(), 'order'];
  }
  private _orderMutationOptions(
    options?: UseMutationOptions<
      Order201Response,
      OrderError,
      OrderOperationRequest
    >,
  ): UseMutationOptions<Order201Response, OrderError, OrderOperationRequest> {
    return {
      mutationFn: (input) => this.$client.order(input),
      mutationKey: this._orderMutationKey(),
      ...options,
    };
  }
  private _userMutationKey() {
    return [...this.queryKey(), 'user'];
  }
  private _userMutationOptions(
    options?: UseMutationOptions<
      User201Response,
      UserError,
      UserOperationRequest
    >,
  ): UseMutationOptions<User201Response, UserError, UserOperationRequest> {
    return {
      mutationFn: (input) => this.$client.user(input),
      mutationKey: this._userMutationKey(),
      ...options,
    };
  }

  public order = {
    mutationKey: this._orderMutationKey.bind(this),
    mutationOptions: this._orderMutationOptions.bind(this),
  };
  public user = {
    mutationKey: this._userMutationKey.bind(this),
    mutationOptions: this._userMutationOptions.bind(this),
  };
}
"
`;

exports[`openApiTsHooksGenerator - Duplicate Types > should handle mutation operations with duplicate type names 1`] = `
"import type { UseMutationOptions } from '@tanstack/react-query';
import type {
  UpdateOperationRequest,
  UpdateError,
  Update200Response,
} from './types.gen.js';
import { TestApi } from './client.gen.js';

export interface TestApiOptionsProxyConfig {
  client: TestApi;
}

export class TestApiOptionsProxy {
  private $client: TestApi;

  constructor({ client }: TestApiOptionsProxyConfig) {
    this.$client = client;

    this._updateMutationKey = this._updateMutationKey.bind(this);
    this._updateMutationOptions = this._updateMutationOptions.bind(this);
  }

  public queryKey = () => ['TestApi'];

  private _updateMutationKey() {
    return [...this.queryKey(), 'update'];
  }
  private _updateMutationOptions(
    options?: UseMutationOptions<
      Update200Response,
      UpdateError,
      UpdateOperationRequest
    >,
  ): UseMutationOptions<
    Update200Response,
    UpdateError,
    UpdateOperationRequest
  > {
    return {
      mutationFn: (input) => this.$client.update(input),
      mutationKey: this._updateMutationKey(),
      ...options,
    };
  }

  public update = {
    mutationKey: this._updateMutationKey.bind(this),
    mutationOptions: this._updateMutationOptions.bind(this),
  };
}
"
`;

exports[`openApiTsHooksGenerator - Duplicate Types > should handle query operations with duplicate type names 1`] = `
"import type { QueryFilters, UseQueryOptions } from '@tanstack/react-query';
import type {
  SearchOperationRequest,
  SearchError,
  Search200Response,
} from './types.gen.js';
import { TestApi } from './client.gen.js';

export interface TestApiOptionsProxyConfig {
  client: TestApi;
}

export class TestApiOptionsProxy {
  private $client: TestApi;

  constructor({ client }: TestApiOptionsProxyConfig) {
    this.$client = client;

    this._searchQueryKey = this._searchQueryKey.bind(this);
    this._searchQueryOptions = this._searchQueryOptions.bind(this);
    this._searchQueryFilter = this._searchQueryFilter.bind(this);
  }

  public queryKey = () => ['TestApi'];

  private _searchQueryKey(input: SearchOperationRequest) {
    return [...this.queryKey(), 'search', input];
  }
  private _searchQueryOptions(
    input: SearchOperationRequest,
    options?: Omit<
      UseQueryOptions<Search200Response, SearchError>,
      'queryFn' | 'queryKey'
    > &
      Partial<
        Pick<
          UseQueryOptions<Search200Response, SearchError>,
          'queryFn' | 'queryKey'
        >
      >,
  ): UseQueryOptions<Search200Response, SearchError> {
    return {
      queryFn: () => this.$client.search(input),
      queryKey: this._searchQueryKey(input),
      ...options,
    };
  }
  private _searchQueryFilter(
    input: SearchOperationRequest,
    filter?: QueryFilters,
  ): QueryFilters {
    return {
      queryKey: this._searchQueryKey(input),
      ...filter,
    };
  }

  public search = {
    queryKey: this._searchQueryKey.bind(this),
    queryOptions: this._searchQueryOptions.bind(this),
    queryFilter: this._searchQueryFilter.bind(this),
  };
}
"
`;
